File: backend/app/domain/spatial_analysis.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from pydantic import Field, model_validator

from app.domain.analytics import (
    AnalyticsCapability,
    AnalyticsContextRequest,
    AnalyticsDatasetContext,
    AnalyticsFilterSummary,
    AnalyticsRuleContext,
)


class SpatialAnalysisMode(StrEnum):
    BIN_MAP = "BIN_MAP"
    PARAMETER_HEATMAP = "PARAMETER_HEATMAP"
    PARAMETER_FAIL_OVERLAY = "PARAMETER_FAIL_OVERLAY"
    COMPOSITE_FAILURE = "COMPOSITE_FAILURE"
    ZONE_COMPARISON = "ZONE_COMPARISON"


class SpatialAnalysisRequest(AnalyticsContextRequest):
    mode: SpatialAnalysisMode
    focus_dataset_id: int | None = Field(default=None, gt=0)
    max_points: int = Field(default=20_000, ge=100, le=50_000)
    rule_code: str | None = Field(default=None, pattern=r"^[A-Z][A-Z0-9_]{2,127}$")
    rule_version: str | None = Field(
        default=None, pattern=r"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$"
    )

    @model_validator(mode="after")
    def mode_contract_is_complete(self) -> SpatialAnalysisRequest:
        if self.focus_dataset_id is not None and self.focus_dataset_id not in {
            item.dataset_id for item in self.datasets
        }:
            raise ValueError("focus_dataset_id must belong to the selected context")
        parameter_modes = {
            SpatialAnalysisMode.PARAMETER_HEATMAP,
            SpatialAnalysisMode.PARAMETER_FAIL_OVERLAY,
        }
        if self.mode in parameter_modes and len(self.parameters) != 1:
            raise ValueError("parameter spatial modes require exactly one parameter")
        if (
            self.mode
            in {
                SpatialAnalysisMode.BIN_MAP,
                SpatialAnalysisMode.COMPOSITE_FAILURE,
            }
            and self.parameters
        ):
            raise ValueError("this spatial mode does not accept parameters")
        if self.mode == SpatialAnalysisMode.ZONE_COMPARISON:
            if len(self.parameters) > 1:
                raise ValueError("zone comparison accepts at most one parameter")
            if self.rule_code is None or self.rule_version is None:
                raise ValueError("zone comparison requires an approved rule reference")
        elif self.rule_code is not None or self.rule_version is not None:
            raise ValueError("rule reference is only accepted by a gated spatial mode")
        return self


@dataclass(frozen=True, slots=True)
class SpatialColorDomain:
    minimum: float
    maximum: float
    p02: float
    p98: float


@dataclass(frozen=True, slots=True)
class SpatialPoint:
    dataset_id: int | None
    version_no: int | None
    lot_id: str | None
    wafer_id: str | None
    x: int
    y: int
    bin_code: str | None
    result: str | None
    value: float | None
    unit: str | None
    lsl: float | None
    usl: float | None
    spec_status: str | None
    drilldown_key: str | None
    observed_count: int
    fail_count: int
    fail_ratio: float | None
    wafer_count: int
    zone: str | None = None
    raw_bin_code: str | None = None
    bin_mapping_set_id: int | None = None
    bin_mapping_version: str | None = None
    bin_name: str | None = None
    failure_mode: str | None = None
    bin_is_pass: bool | None = None
    spec_set_id: int | None = None
    spec_version: str | None = None
    quadrant: str | None = None
    member_drilldown_keys: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class SpatialWaferIdentity:
    key: str
    dataset_id: int
    version_no: int
    lot_id: str
    wafer_id: str


@dataclass(frozen=True, slots=True)
class SpatialZoneSummary:
    zone: str
    unit_count: int
    pass_count: int
    fail_count: int
    unknown_count: int
    yield_rate: float | None
    measured_count: int
    missing_measurement_count: int
    mean: float | None
    minimum: float | None
    maximum: float | None
    drilldown_key: str | None = None
    member_drilldown_keys: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class SpatialQuadrantSummary:
    quadrant: str
    unit_count: int
    pass_count: int
    fail_count: int
    unknown_count: int
    yield_rate: float | None
    measured_count: int
    missing_measurement_count: int
    mean: float | None
    minimum: float | None
    maximum: float | None
    member_drilldown_keys: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class SpatialZoneGeometry:
    center_x: float
    center_y: float
    radius: float
    center_ratio: float
    mid_ratio: float
    quadrant_axis_rotation_degrees: float
    quadrant_y_direction: str
    quadrant_labels_ccw: tuple[str, str, str, str]


@dataclass(frozen=True, slots=True)
class SpatialDataQuality:
    input_units: int
    returned_points: int
    wafer_count: int
    missing_coordinate_count: int
    duplicate_coordinate_count: int
    measured_count: int
    missing_measurement_count: int
    layer_point_count: int


@dataclass(frozen=True, slots=True)
class SpatialAnalysisResult:
    contract_version: str
    dataset_context: AnalyticsDatasetContext
    filter_summary: AnalyticsFilterSummary
    rule_context: AnalyticsRuleContext
    capabilities: tuple[AnalyticsCapability, ...]
    mode: str
    parameter: str | None
    color_domain: SpatialColorDomain | None
    data_quality: SpatialDataQuality
    points: tuple[SpatialPoint, ...]
    wafer_manifest: tuple[SpatialWaferIdentity, ...]
    wafer_layers: tuple[SpatialPoint, ...]
    zones: tuple[SpatialZoneSummary, ...]
    warnings: tuple[str, ...]
    computed_at: str
    zone_geometry: SpatialZoneGeometry | None = None
    quadrants: tuple[SpatialQuadrantSummary, ...] = ()

File: frontend/src/api/spatialAnalysis.ts
import type {
  AnalyticsCapability,
  AnalyticsContextRequest,
  AnalyticsDatasetContext,
  AnalyticsFilterSummary,
  AnalyticsRuleContext,
} from "./analytics";
import { apiRequest } from "./auth";

export type SpatialAnalysisMode = "BIN_MAP" | "PARAMETER_HEATMAP" | "PARAMETER_FAIL_OVERLAY" | "COMPOSITE_FAILURE" | "ZONE_COMPARISON";

export interface SpatialAnalysisRequest extends AnalyticsContextRequest {
  mode: SpatialAnalysisMode;
  focus_dataset_id?: number | null;
  max_points: number;
  rule_code?: string | null;
  rule_version?: string | null;
}

export interface SpatialColorDomain {
  minimum: number;
  maximum: number;
  p02: number;
  p98: number;
}

export interface SpatialPoint {
  dataset_id: number | null;
  version_no: number | null;
  lot_id: string | null;
  wafer_id: string | null;
  x: number;
  y: number;
  bin_code: string | null;
  result: string | null;
  value: number | null;
  unit: string | null;
  lsl: number | null;
  usl: number | null;
  spec_status: string | null;
  drilldown_key: string | null;
  observed_count: number;
  fail_count: number;
  fail_ratio: number | null;
  wafer_count: number;
  zone?: string | null;
  raw_bin_code?: string | null;
  bin_mapping_set_id?: number | null;
  bin_mapping_version?: string | null;
  bin_name?: string | null;
  failure_mode?: string | null;
  bin_is_pass?: boolean | null;
  spec_set_id?: number | null;
  spec_version?: string | null;
  quadrant?: string | null;
  member_drilldown_keys?: string[];
}

export interface SpatialWaferIdentity {
  key: string;
  dataset_id: number;
  version_no: number;
  lot_id: string;
  wafer_id: string;
}

export interface SpatialZoneSummary {
  zone: string;
  unit_count: number;
  pass_count: number;
  fail_count: number;
  unknown_count: number;
  yield_rate: number | null;
  measured_count: number;
  missing_measurement_count: number;
  mean: number | null;
  minimum: number | null;
  maximum: number | null;
  drilldown_key?: string | null;
  member_drilldown_keys?: string[];
}

export interface SpatialQuadrantSummary {
  quadrant: string;
  unit_count: number;
  pass_count: number;
  fail_count: number;
  unknown_count: number;
  yield_rate: number | null;
  measured_count: number;
  missing_measurement_count: number;
  mean: number | null;
  minimum: number | null;
  maximum: number | null;
  member_drilldown_keys: string[];
}

export interface SpatialZoneGeometry {
  center_x: number;
  center_y: number;
  radius: number;
  center_ratio: number;
  mid_ratio: number;
  quadrant_axis_rotation_degrees: number;
  quadrant_y_direction: "UP" | "DOWN";
  quadrant_labels_ccw: [string, string, string, string];
}

export interface SpatialDataQuality {
  input_units: number;
  returned_points: number;
  wafer_count: number;
  missing_coordinate_count: number;
  duplicate_coordinate_count: number;
  measured_count: number;
  missing_measurement_count: number;
  layer_point_count: number;
}

export interface SpatialAnalysisResult {
  contract_version: string;
  dataset_context: AnalyticsDatasetContext;
  filter_summary: AnalyticsFilterSummary;
  rule_context: AnalyticsRuleContext;
  capabilities: AnalyticsCapability[];
  mode: string;
  parameter: string | null;
  color_domain: SpatialColorDomain | null;
  data_quality: SpatialDataQuality;
  points: SpatialPoint[];
  wafer_manifest: SpatialWaferIdentity[];
  wafer_layers: SpatialPoint[];
  zones: SpatialZoneSummary[];
  warnings: string[];
  computed_at: string;
  zone_geometry?: SpatialZoneGeometry | null;
  quadrants?: SpatialQuadrantSummary[];
}

export function analyzeSpatial(request: SpatialAnalysisRequest): Promise<SpatialAnalysisResult> {
  return apiRequest("/api/v1/analytics/spatial", {
    method: "POST",
    body: JSON.stringify(request),
  });
}
